flask/cli.py

```diff
@@ -37,19 +37,46 @@ def find_best_app(module):
     # Search for the most common names first.
     for attr_name in 'app', 'application':
         app = getattr(module, attr_name, None)
-        if app is not None and isinstance(app, Flask):
+        if isinstance(app, Flask):
             return app
 
     # Otherwise find the only object that is a Flask instance.
-    matches = [v for k, v in iteritems(module.__dict__)
-               if isinstance(v, Flask)]
+    matches = [
+        v for k, v in iteritems(module.__dict__) if isinstance(v, Flask)
+    ]
 
     if len(matches) == 1:
         return matches[0]
-    raise NoAppException('Failed to find application in module "%s".  Are '
-                         'you sure it contains a Flask application?  Maybe '
-                         'you wrapped it in a WSGI middleware or you are '
-                         'using a factory function.' % module.__name__)
+    elif len(matches) > 1:
+        raise NoAppException(
+            'Auto-detected multiple Flask applications in module "{module}".'
+            ' Use "FLASK_APP={module}:name" to specify the correct'
+            ' one.'.format(module=module.__name__)
+        )
+
+    # Search for app factory callables.
+    for attr_name in 'create_app', 'make_app':
+        app_factory = getattr(module, attr_name, None)
+
+        if callable(app_factory):
+            try:
+                app = app_factory()
+
```

Comment:
The intention is that the factory function could be built in a way that it requires positional arguments (compare http://flask.pocoo.org/docs/0.12/patterns/appfactories/) and thus the call could fail with a TypeError. I'd rather throw a custom exception in that case.